// @sun-asterisk/sunlint/rules/common/C002_no_duplicate_code/analyzer.js
/**
 * C002 - No Duplicate Code Analyzer (AST-based with ts-morph)
 * 
 * This analyzer uses TypeScript Compiler API via ts-morph for accurate
 * semantic analysis and duplicate detection.
 * 
 * Advantages over regex-based approach:
 * - 95-99% accuracy vs 70-80%
 * - True semantic understanding via AST
 * - Precise function/class extraction
 * - Better handling of edge cases
 * 
 * Trade-offs:
 * - TypeScript/JavaScript only (no multi-language support)
 * - Slightly slower (~30% more time)
 * - Requires ts-morph dependency
 */

const fs = require('fs');
const path = require('path');
const { Project, SyntaxKind } = require('ts-morph');

class C002AnalyzerAST {
    constructor(config = {}) {
        this.config = {
            minLines: config.minLines || 10,
            // Stricter threshold (95%) to detect only true monkey coding (copy-paste)
            // This filters out intentional boilerplate patterns (87-90% similarity)
            similarityThreshold: config.similarityThreshold || 0.95,
        };
    }

    /**
     * Analyze files for duplicate code
     * @param {string[]} filePaths - Array of file paths to analyze
     * @param {string} language - Language (typescript/javascript)
     * @returns {Array} Array of violations
     */
    analyze(filePaths, language = 'typescript') {
        try {
            // Filter TypeScript/JavaScript files only
            const validFiles = filePaths.filter(fp => 
                /\.(ts|tsx|js|jsx)$/i.test(fp)
            );

            if (validFiles.length === 0) {
                return [];
            }

            // Create fresh project instance for each analyze() call
            // This prevents file caching issues
            this.project = new Project({
                compilerOptions: {
                    target: 99, // ESNext
                    allowJs: true,
                },
                skipAddingFilesFromTsConfig: true,
            });

            // Add files to ts-morph project (in batches for performance)
            const sourceFiles = [];
            const batchSize = 10;
            
            for (let i = 0; i < validFiles.length; i += batchSize) {
                const batch = validFiles.slice(i, i + batchSize);
                
                batch.forEach(fp => {
                    try {
                        const sourceFile = this.project.addSourceFileAtPath(fp);
                        sourceFiles.push(sourceFile);
                    } catch (error) {
                        // Silently skip unparseable files
                    }
                });
            }

            if (sourceFiles.length === 0) {
                return [];
            }

            // Extract code blocks from all files
            const allBlocks = [];
            sourceFiles.forEach(sourceFile => {
                const blocks = this.extractCodeBlocks(sourceFile);
                allBlocks.push(...blocks);
            });

            // Find duplicates
            const duplicateGroups = this.findDuplicates(allBlocks);

            // Create violations
            const violations = this.createViolations(duplicateGroups);

            return violations;

        } catch (error) {
            console.error('Error analyzing files with C002-AST:', error.message);
            return [];
        }
    }

    /**
     * Extract code blocks from source file using AST
     * @param {SourceFile} sourceFile - ts-morph source file
     * @returns {Array} Array of code blocks
     */
    extractCodeBlocks(sourceFile) {
        const blocks = [];
        const filePath = sourceFile.getFilePath();

        // Extract functions
        const functions = sourceFile.getFunctions();
        functions.forEach(fn => {
            const block = this.createBlockFromNode(fn, 'function', filePath);
            if (block && block.nonCommentLines >= this.config.minLines) {
                blocks.push(block);
            }
        });

        // Extract arrow functions and function expressions
        const variableDeclarations = sourceFile.getVariableDeclarations();
        variableDeclarations.forEach(varDecl => {
            const initializer = varDecl.getInitializer();
            if (initializer) {
                const kind = initializer.getKind();
                if (kind === SyntaxKind.ArrowFunction || 
                    kind === SyntaxKind.FunctionExpression) {
                    const block = this.createBlockFromNode(initializer, 'arrow-function', filePath, varDecl.getName());
                    if (block && block.nonCommentLines >= this.config.minLines) {
                        blocks.push(block);
                    }
                }
            }
        });

        // Extract classes
        const classes = sourceFile.getClasses();
        classes.forEach(cls => {
            const block = this.createBlockFromNode(cls, 'class', filePath);
            if (block && block.nonCommentLines >= this.config.minLines) {
                blocks.push(block);
            }

            // Extract methods from class
            const methods = cls.getMethods();
            methods.forEach(method => {
                const block = this.createBlockFromNode(method, 'method', filePath);
                if (block && block.nonCommentLines >= this.config.minLines) {
                    blocks.push(block);
                }
            });
        });

        // Extract interfaces
        const interfaces = sourceFile.getInterfaces();
        interfaces.forEach(iface => {
            const block = this.createBlockFromNode(iface, 'interface', filePath);
            if (block && block.nonCommentLines >= this.config.minLines) {
                blocks.push(block);
            }
        });

        return blocks;
    }

    /**
     * Create code block from AST node
     * @param {Node} node - AST node
     * @param {string} type - Block type
     * @param {string} filePath - File path
     * @param {string} customName - Custom name (for arrow functions)
     * @returns {Object} Code block
     */
    createBlockFromNode(node, type, filePath, customName = null) {
        const startLine = node.getStartLineNumber();
        const endLine = node.getEndLineNumber();
        const fullText = node.getText();
        
        // Count non-comment lines
        const lines = fullText.split('\n');
        const nonCommentLines = this.countNonCommentLines(lines);
        
        if (nonCommentLines < this.config.minLines) {
            return null;
        }

        // Get name
        let name = customName;
        if (!name && typeof node.getName === 'function') {
            name = node.getName();
        }
        if (!name) {
            name = 'anonymous';
        }

        // Get normalized code for comparison
        const normalizedCode = this.normalizeCode(fullText);

        // Extract semantic tokens (identifiers)
        const tokens = this.extractTokensFromNode(node);

        // Extract structure (skeleton)
        const structure = this.extractStructureFromNode(node);

        return {
            startLine,
            endLine,
            lineCount: endLine - startLine + 1,
            nonCommentLines,
            filePath,
            type,
            name,
            fullText,
            normalizedCode,
            tokens,
            structure,
            node // Keep reference to AST node
        };
    }

    /**
     * Count non-comment lines in code
     * @param {string[]} lines - Array of lines
     * @returns {number} Count
     */
    countNonCommentLines(lines) {
        let count = 0;
        let inBlockComment = false;

        for (const line of lines) {
            const trimmed = line.trim();
            
            // Check block comment start/end
            if (trimmed.startsWith('/*')) inBlockComment = true;
            if (inBlockComment) {
                if (trimmed.endsWith('*/')) inBlockComment = false;
                continue;
            }
            
            // Skip single-line comments and empty lines
            if (trimmed.startsWith('//') || trimmed.length === 0) {
                continue;
            }
            
            count++;
        }

        return count;
    }

    /**
     * Extract tokens (identifiers) from AST node (optimized)
     * @param {Node} node - AST node
     * @returns {Array<string>} Array of tokens
     */
    extractTokensFromNode(node) {
        const tokens = [];
        const identifiers = node.getDescendantsOfKind(SyntaxKind.Identifier);
        
        // Limit to first 100 identifiers for performance
        const limit = Math.min(identifiers.length, 100);
        
        for (let i = 0; i < limit; i++) {
            const text = identifiers[i].getText();
            // Skip TypeScript/JavaScript keywords
            if (!this.isKeyword(text) && text.length > 1) {
                tokens.push(text);
            }
        }

        return tokens;
    }

    /**
     * Check if string is a keyword
     * @param {string} text - Text to check
     * @returns {boolean} True if keyword
     */
    isKeyword(text) {
        const keywords = new Set([
            'function', 'const', 'let', 'var', 'if', 'else', 'return', 'for', 'while',
            'class', 'interface', 'extends', 'implements', 'import', 'export', 'from',
            'async', 'await', 'try', 'catch', 'throw', 'new', 'this', 'super',
            'true', 'false', 'null', 'undefined', 'typeof', 'instanceof', 'of', 'in'
        ]);
        return keywords.has(text);
    }

    /**
     * Extract structure (skeleton) from AST node
     * @param {Node} node - AST node
     * @returns {string} Structure string
     */
    extractStructureFromNode(node) {
        // Get syntax kind path for structure comparison (optimized)
        const structure = [];
        let depth = 0;
        const maxDepth = 20; // Limit depth for performance
        
        node.forEachDescendant((descendant, traversal) => {
            depth++;
            
            // Stop if too deep or too many nodes
            if (depth > maxDepth || structure.length > 200) {
                traversal.stop();
                return;
            }
            
            const kind = descendant.getKindName();
            structure.push(kind);
        });

        return structure.join('|');
    }

    /**
     * Normalize code for comparison
     * @param {string} code - Raw code
     * @returns {string} Normalized code
     */
    normalizeCode(code) {
        let normalized = code;

        if (this.config.ignoreComments) {
            normalized = normalized.replace(/\/\/.*$/gm, '');
            normalized = normalized.replace(/\/\*[\s\S]*?\*\//g, '');
        }

        if (this.config.ignoreWhitespace) {
            normalized = normalized
                .replace(/\s+/g, ' ')
                .replace(/\s*{\s*/g, '{')
                .replace(/\s*}\s*/g, '}')
                .replace(/\s*;\s*/g, ';')
                .replace(/\s*,\s*/g, ',')
                .trim();
        }

        if (this.config.ignoreEmptyLines) {
            normalized = normalized
                .split('\n')
                .filter(line => line.trim().length > 0)
                .join('\n');
        }

        return normalized;
    }

    /**
     * Find duplicate code blocks using AST-based semantic analysis
     * @param {Array} blocks - Array of code blocks
     * @returns {Array} Array of duplicate groups
     */
    findDuplicates(blocks) {
        const duplicateGroups = [];
        const processedBlocks = new Set();

        for (let i = 0; i < blocks.length; i++) {
            if (processedBlocks.has(i)) continue;

            const currentBlock = blocks[i];
            const duplicates = [currentBlock];

            for (let j = i + 1; j < blocks.length; j++) {
                if (processedBlocks.has(j)) continue;

                const otherBlock = blocks[j];

                // Calculate similarity
                const similarity = this.calculateSimilarity(currentBlock, otherBlock);

                if (similarity >= this.config.similarityThreshold) {
                    // Check if this is an intentional pattern
                    if (this.isIntentionalPattern(currentBlock, otherBlock)) {
                        continue;
                    }

                    duplicates.push(otherBlock);
                    processedBlocks.add(j);
                }
            }

            if (duplicates.length > 1) {
                duplicateGroups.push(duplicates);
                processedBlocks.add(i);
            }
        }

        return duplicateGroups;
    }

    /**
     * Calculate similarity between two blocks using AST (optimized)
     * @param {Object} block1 - First block
     * @param {Object} block2 - Second block
     * @returns {number} Similarity (0-1)
     */
    calculateSimilarity(block1, block2) {
        // Quick rejection: if structures are very different, skip expensive comparison
        if (Math.abs(block1.structure.length - block2.structure.length) > 100) {
            return 0;
        }
        
        // Use structure similarity as primary metric
        const structureSim = this.calculateStructureSimilarity(block1.structure, block2.structure);
        
        // Early exit if structure is too different
        if (structureSim < 0.5) {
            return structureSim * 0.6; // Don't waste time on code comparison
        }
        
        // Use normalized code as secondary metric only if structure is similar
        const codeSim = this.calculateStringSimilarity(
            block1.normalizedCode.substring(0, 500), // Limit string length
            block2.normalizedCode.substring(0, 500)
        );
        
        // Weighted average: 60% structure, 40% code
        return structureSim * 0.6 + codeSim * 0.4;
    }

    /**
     * Calculate structure similarity
     * @param {string} structure1 - First structure
     * @param {string} structure2 - Second structure
     * @returns {number} Similarity (0-1)
     */
    calculateStructureSimilarity(structure1, structure2) {
        if (structure1 === structure2) return 1.0;
        
        const tokens1 = structure1.split('|');
        const tokens2 = structure2.split('|');
        
        const maxLen = Math.max(tokens1.length, tokens2.length);
        if (maxLen === 0) return 1.0;
        
        let matches = 0;
        const minLen = Math.min(tokens1.length, tokens2.length);
        
        for (let i = 0; i < minLen; i++) {
            if (tokens1[i] === tokens2[i]) matches++;
        }
        
        return matches / maxLen;
    }

    /**
     * Calculate string similarity using Levenshtein distance
     * @param {string} str1 - First string
     * @param {string} str2 - Second string
     * @returns {number} Similarity (0-1)
     */
    calculateStringSimilarity(str1, str2) {
        if (str1 === str2) return 1.0;

        const longer = str1.length > str2.length ? str1 : str2;
        const shorter = str1.length > str2.length ? str2 : str1;

        if (longer.length === 0) return 1.0;

        const distance = this.levenshteinDistance(longer, shorter);
        return (longer.length - distance) / longer.length;
    }

    /**
     * Calculate Levenshtein distance
     * @param {string} str1 - First string
     * @param {string} str2 - Second string
     * @returns {number} Distance
     */
    levenshteinDistance(str1, str2) {
        const matrix = Array(str2.length + 1).fill().map(() => Array(str1.length + 1).fill(0));

        for (let i = 0; i <= str1.length; i++) matrix[0][i] = i;
        for (let j = 0; j <= str2.length; j++) matrix[j][0] = j;

        for (let j = 1; j <= str2.length; j++) {
            for (let i = 1; i <= str1.length; i++) {
                const cost = str1[i - 1] === str2[j - 1] ? 0 : 1;
                matrix[j][i] = Math.min(
                    matrix[j][i - 1] + 1,
                    matrix[j - 1][i] + 1,
                    matrix[j - 1][i - 1] + cost
                );
            }
        }

        return matrix[str2.length][str1.length];
    }

    /**
     * Check if two blocks are intentional patterns using AST analysis
     * @param {Object} block1 - First block
     * @param {Object} block2 - Second block
     * @returns {boolean} True if intentional pattern
     */
    isIntentionalPattern(block1, block2) {
        // 0. Skip if same location (should not happen, but safety check)
        if (block1.filePath === block2.filePath &&
            block1.startLine === block2.startLine &&
            block1.endLine === block2.endLine) {
            return true; // Treat as intentional to skip
        }

        // 1. Simple JSX/HTML wrapper detection (no business logic)
        // Pattern: function Component({ ...props }) { return <span {...props} /> }
        if (this.isSimpleJSXWrapper(block1.fullText) && 
            this.isSimpleJSXWrapper(block2.fullText)) {
            
            // If function names are different, it's intentional pattern
            if (block1.name !== block2.name && 
                block1.name !== 'anonymous' && 
                block2.name !== 'anonymous') {
                return true;
            }
        }
        
        // 2. React component wrapper detection (library wrappers)
        // Pattern: function Component({ ...props }) { return <Primitive.X {...props} /> }
        if (this.isReactWrapperComponent(block1.fullText) && 
            this.isReactWrapperComponent(block2.fullText)) {
            
            // Extract wrapped component names
            const wrapped1 = this.extractWrappedComponentName(block1.fullText);
            const wrapped2 = this.extractWrappedComponentName(block2.fullText);
            
            // If wrapping different components, it's intentional
            if (wrapped1 && wrapped2 && wrapped1 !== wrapped2) {
                return true;
            }
        }

        // 3. Token-based analysis
        const tokenSimilarity = this.calculateTokenSimilarity(block1.tokens, block2.tokens);
        const structureSimilarity = this.calculateStructureSimilarity(block1.structure, block2.structure);

        // High structure similarity + Low token similarity = Intentional pattern
        if (structureSimilarity >= 0.90 && tokenSimilarity <= 0.65) {
            return true;
        }

        // 4. Unique token ratio
        const uniqueTokens1 = block1.tokens.filter(t => !block2.tokens.includes(t));
        const uniqueTokens2 = block2.tokens.filter(t => !block1.tokens.includes(t));
        const totalTokens = Math.max(block1.tokens.length, block2.tokens.length);

        if (totalTokens > 0) {
            const uniqueRatio = (uniqueTokens1.length + uniqueTokens2.length) / (2 * totalTokens);

            // 10-40% different tokens + high structure similarity = Intentional
            if (uniqueRatio >= 0.10 && uniqueRatio <= 0.40 && structureSimilarity >= 0.85) {
                return true;
            }
        }

        // 5. Function name completely different = intentional pattern
        if (block1.name !== 'anonymous' && block2.name !== 'anonymous' && block1.name !== block2.name) {
            const nameSimilarity = this.calculateStringSimilarity(
                block1.name.toLowerCase(),
                block2.name.toLowerCase()
            );

            // Names less than 50% similar = completely different purposes
            if (nameSimilarity < 0.50) {
                return true;
            }
        }

        return false;
    }

    /**
     * Check if code is a simple JSX/HTML wrapper (no business logic)
     * @param {string} code - Full code text
     * @returns {boolean} True if it's a simple wrapper
     */
    isSimpleJSXWrapper(code) {
        // Pattern: function that just returns JSX with props spreading, no logic
        const lines = code.split('\n').filter(l => l.trim() && !l.trim().startsWith('//')).length;
        
        // Simple wrappers are typically 5-20 lines
        if (lines < 5 || lines > 20) {
            return false;
        }
        
        // Must have {...props} spread
        if (!code.includes('{...props}')) {
            return false;
        }
        
        // Must NOT have business logic
        const hasLogic = /\b(if|else|switch|for|while|map|filter|reduce|forEach)\b/.test(code);
        if (hasLogic) {
            return false;
        }
        
        // Must have single return statement with JSX
        const returnCount = (code.match(/\breturn\b/g) || []).length;
        if (returnCount !== 1) {
            return false;
        }
        
        // Must return JSX element (either component or HTML element)
        const hasJSXReturn = /<[a-zA-Z]/.test(code);
        return hasJSXReturn;
    }

    /**
     * Check if code is a React component wrapper
     * @param {string} code - Full code text
     * @returns {boolean} True if it's a wrapper
     */
    isReactWrapperComponent(code) {
        // Pattern: very short function that just returns <Component {...props} />
        const lines = code.split('\n').filter(l => l.trim() && !l.trim().startsWith('//')).length;
        
        // Wrapper components are typically 5-15 lines
        if (lines < 5 || lines > 20) {
            return false;
        }
        
        // Must have {...props} spread
        if (!code.includes('{...props}')) {
            return false;
        }
        
        // Must return JSX with a component (starts with uppercase)
        const returnMatch = code.match(/return\s*\(?<([A-Z][a-zA-Z0-9.]*)/);
        return returnMatch !== null;
    }

    /**
     * Extract wrapped component name from wrapper code
     * @param {string} code - Full code text
     * @returns {string|null} Component name or null
     */
    extractWrappedComponentName(code) {
        // Match: return <ComponentName or return <Component.SubComponent
        const match = code.match(/return\s*\(?<([A-Z][a-zA-Z0-9.]*)/);
        return match ? match[1] : null;
    }

    /**
     * Calculate token similarity (Jaccard index)
     * @param {Array<string>} tokens1 - First token array
     * @param {Array<string>} tokens2 - Second token array
     * @returns {number} Similarity (0-1)
     */
    calculateTokenSimilarity(tokens1, tokens2) {
        if (tokens1.length === 0 && tokens2.length === 0) return 1.0;
        if (tokens1.length === 0 || tokens2.length === 0) return 0.0;

        const set1 = new Set(tokens1);
        const set2 = new Set(tokens2);

        const intersection = new Set([...set1].filter(x => set2.has(x)));
        const union = new Set([...set1, ...set2]);

        return intersection.size / union.size;
    }

    /**
     * Create violations from duplicate groups
     * @param {Array} duplicateGroups - Array of duplicate groups
     * @returns {Array} Array of violations
     */
    createViolations(duplicateGroups) {
        const violations = [];

        duplicateGroups.forEach(group => {
            const firstBlock = group[0];
            
            // Use relative path from project root to avoid confusion
            const locations = group.map(block => {
                // Try to get relative path, fallback to absolute if not possible
                let displayPath = block.filePath;
                
                // Find common project root (where package.json or node_modules exists)
                const parts = block.filePath.split(path.sep);
                const projectRootIndex = parts.findIndex(p => p === 'components' || p === 'hooks' || p === 'lib' || p === 'src' || p === 'app');
                
                if (projectRootIndex > 0) {
                    displayPath = parts.slice(projectRootIndex).join('/');
                } else {
                    // Fallback: show last 2-3 path segments
                    displayPath = parts.slice(-3).join('/');
                }
                
                return `${displayPath}:${block.startLine}-${block.endLine}`;
            });

            // Generate suggestions based on context
            const suggestions = this.generateRefactoringSuggestions(group);

            const violation = {
                ruleId: 'C002',  // Changed from 'rule' to 'ruleId' for consistency with output-service.js
                severity: 'warning',
                message: `Duplicate ${firstBlock.type} detected (${firstBlock.nonCommentLines} non-comment lines). ${suggestions[0]} Found in ${group.length} locations: ${locations.join(', ')}`,
                file: firstBlock.filePath,
                line: firstBlock.startLine,
                column: 1,
                endLine: firstBlock.endLine,
                data: {
                    duplicateCount: group.length,
                    nonCommentLines: firstBlock.nonCommentLines,
                    locations: locations,
                    blockType: firstBlock.type,
                    suggestions: suggestions
                }
            };

            violations.push(violation);
        });

        return violations;
    }

    /**
     * Generate refactoring suggestions based on duplicate context
     * @param {Array} group - Duplicate group
     * @returns {Array<string>} Array of suggestions
     */
    generateRefactoringSuggestions(group) {
        const suggestions = [];
        const firstBlock = group[0];
        const files = new Set(group.map(b => b.filePath));

        if (files.size === 1) {
            // Same file duplicates
            suggestions.push('Extract into a shared utility function in this file');
        } else {
            // Cross-file duplicates
            suggestions.push('Extract into a shared utility module or helper file');
        }

        // Type-specific suggestions
        if (firstBlock.type === 'function' || firstBlock.type === 'arrow-function') {
            suggestions.push('Consider using function composition or higher-order functions');
        } else if (firstBlock.type === 'class') {
            suggestions.push('Use inheritance or composition to share common behavior');
        } else if (firstBlock.type === 'method') {
            suggestions.push('Extract common logic into a base class or mixin');
        }

        return suggestions;
    }
}

// Export both class and SunLint-compatible API
module.exports = C002AnalyzerAST;

// SunLint integration: Export check() method
module.exports.check = async function check(filePaths, language) {
    console.log(`🔍 C002 Analyzer called with ${filePaths.length} files, language: ${language}`);
    const analyzer = new C002AnalyzerAST();
    const violations = await analyzer.analyze(filePaths);
    console.log(`✅ C002 Analyzer found ${violations.length} violations`);
    return violations;
};
