// @sun-asterisk/sunlint/rules/security/S005_no_origin_auth/analyzer.js
/**
 * S005 - No Origin Header Authentication
 *
 * Main analyzer using symbol-based analysis to detect use of Origin header
 * for authentication or authorization decisions.
 *
 * Based on:
 * - OWASP A07:2021 - Identification and Authentication Failures
 * - CWE-290: Authentication Bypass by Spoofing
 */

const S005SymbolBasedAnalyzer = require("./symbol-based-analyzer");

class S005Analyzer {
  constructor(options = {}) {
    this.ruleId = "S005";
    this.semanticEngine = options.semanticEngine || null;
    this.verbose = options.verbose || false;

    try {
      this.symbolAnalyzer = new S005SymbolBasedAnalyzer(this.semanticEngine);
    } catch (e) {
      console.warn(`⚠ [S005] Failed to create symbol analyzer: ${e.message}`);
    }
  }

  async initialize(semanticEngine) {
    this.semanticEngine = semanticEngine;
    if (this.symbolAnalyzer && this.symbolAnalyzer.initialize) {
      await this.symbolAnalyzer.initialize(semanticEngine);
    }
  }

  analyzeSingle(filePath, options = {}) {
    return this.analyze([filePath], "typescript", options);
  }

  async analyze(files, language, options = {}) {
    const violations = [];
    for (const filePath of files) {
      try {
        const vs = await this.analyzeFile(filePath, options);
        violations.push(...vs);
      } catch (e) {
        console.warn(`⚠ [S005] Analysis error for ${filePath}: ${e.message}`);
      }
    }
    return violations;
  }

  async analyzeFile(filePath, options = {}) {
    const violationMap = new Map();

    if (!this.symbolAnalyzer) {
      return [];
    }

    // Skip test files, build directories, and node_modules
    if (this.shouldSkipFile(filePath)) {
      return [];
    }

    try {
      let sourceFile = null;
      if (this.semanticEngine?.project) {
        sourceFile = this.semanticEngine.project.getSourceFile(filePath);
      }

      if (!sourceFile) {
        // Create temporary ts-morph source file
        const fs = require("fs");
        const path = require("path");
        const { Project } = require("ts-morph");
        if (!fs.existsSync(filePath)) {
          throw new Error(`File not found: ${filePath}`);
        }
        const content = fs.readFileSync(filePath, "utf8");
        const tmp = new Project({
          useInMemoryFileSystem: true,
          compilerOptions: { allowJs: true },
        });
        sourceFile = tmp.createSourceFile(path.basename(filePath), content);
      }

      if (sourceFile) {
        const symbolViolations = await this.symbolAnalyzer.analyze(
          sourceFile,
          filePath
        );
        symbolViolations.forEach((v) => {
          const key = `${v.line}:${v.column}:${v.message}`;
          if (!violationMap.has(key)) violationMap.set(key, v);
        });
      }
    } catch (e) {
      console.warn(`⚠ [S005] Symbol analysis failed: ${e.message}`);
    }

    return Array.from(violationMap.values()).map((v) => ({
      ...v,
      filePath,
      file: filePath,
    }));
  }

  shouldSkipFile(filePath) {
    const skipPatterns = [
      "test/",
      "tests/",
      "__tests__/",
      ".test.",
      ".spec.",
      "node_modules/",
      "build/",
      "dist/",
      ".next/",
      "coverage/",
      "vendor/",
      "mocks/",
      ".mock.",
    ];

    return skipPatterns.some((pattern) => filePath.includes(pattern));
  }

  cleanup() {
    if (this.symbolAnalyzer?.cleanup) {
      this.symbolAnalyzer.cleanup();
    }
  }
}

module.exports = S005Analyzer;
