// @sun-asterisk/sunlint/rules/security/S020_no_eval_dynamic_code/analyzer.js
/**
 * S020 Main Analyzer - Avoid using eval() or executing dynamic code
 * Primary: Symbol-based analysis (when available)
 * Fallback: Regex-based for all other cases
 * Command: node cli.js --rule=S020 --input=examples/rule-test-fixtures/rules/S020_no_eval_dynamic_code --engine=heuristic
 */

const S020SymbolBasedAnalyzer = require("./symbol-based-analyzer.js");

class S020Analyzer {
  constructor(options = {}) {
    if (process.env.SUNLINT_DEBUG) {
      console.log(`ðŸ”§ [S020] Constructor called with options:`, !!options);
      console.log(
        `ðŸ”§ [S020] Options type:`,
        typeof options,
        Object.keys(options || {})
      );
    }

    this.ruleId = "S020";
    this.ruleName = "Avoid using eval() or executing dynamic code";
    this.description =
      "Avoid using eval() or executing dynamic code as it can lead to code injection vulnerabilities and compromise application security.";
    this.semanticEngine = options.semanticEngine || null;
    this.verbose = options.verbose || false;

    try {
      this.symbolAnalyzer = new S020SymbolBasedAnalyzer(this.semanticEngine);
      if (process.env.SUNLINT_DEBUG)
        console.log(`ðŸ”§ [S020] Symbol analyzer created successfully`);
    } catch (error) {
      console.error(`ðŸ”§ [S020] Error creating symbol analyzer:`, error);
    }
  }

  async initialize(semanticEngine) {
    this.semanticEngine = semanticEngine;
    if (process.env.SUNLINT_DEBUG)
      console.log(`ðŸ”§ [S020] Main analyzer initializing...`);

    if (this.symbolAnalyzer)
      await this.symbolAnalyzer.initialize?.(semanticEngine);

    if (process.env.SUNLINT_DEBUG)
      console.log(`ðŸ”§ [S020] Main analyzer initialized successfully`);
  }

  analyzeSingle(filePath, options = {}) {
    if (process.env.SUNLINT_DEBUG)
      console.log(`ðŸ” [S020] analyzeSingle() called for: ${filePath}`);
    return this.analyze([filePath], "typescript", options);
  }

  async analyze(files, language, options = {}) {
    if (process.env.SUNLINT_DEBUG) {
      console.log(
        `ðŸ”§ [S020] analyze() method called with ${files.length} files, language: ${language}`
      );
    }

    const violations = [];
    for (const filePath of files) {
      try {
        if (process.env.SUNLINT_DEBUG)
          console.log(`ðŸ”§ [S020] Processing file: ${filePath}`);
        const fileViolations = await this.analyzeFile(filePath, options);
        violations.push(...fileViolations);
        if (process.env.SUNLINT_DEBUG)
          console.log(
            `ðŸ”§ [S020] File ${filePath}: Found ${fileViolations.length} violations`
          );
      } catch (error) {
        console.warn(
          `âš  [S020] Analysis failed for ${filePath}:`,
          error.message
        );
      }
    }

    if (process.env.SUNLINT_DEBUG)
      console.log(`ðŸ”§ [S020] Total violations found: ${violations.length}`);
    return violations;
  }

  async analyzeFile(filePath, options = {}) {
    if (process.env.SUNLINT_DEBUG)
      console.log(`ðŸ” [S020] analyzeFile() called for: ${filePath}`);
    const violationMap = new Map();

    try {
      if (process.env.SUNLINT_DEBUG)
        console.log(`ðŸ”§ [S020] Running symbol-based analysis...`);

      let sourceFile = null;
      if (this.semanticEngine?.project) {
        sourceFile = this.semanticEngine.project.getSourceFile(filePath);
        if (process.env.SUNLINT_DEBUG) {
          console.log(
            `ðŸ”§ [S020] Checked existing semantic engine project: sourceFile=${!!sourceFile}`
          );
        }
      }

      if (!sourceFile) {
        // Create a minimal ts-morph project for this analysis
        if (process.env.SUNLINT_DEBUG)
          console.log(
            `ðŸ”§ [S020] Creating temporary ts-morph project for: ${filePath}`
          );
        try {
          const fs = require("fs");
          const path = require("path");
          const { Project } = require("ts-morph");

          // Check if file exists and read content
          if (!fs.existsSync(filePath)) {
            throw new Error(`File not found: ${filePath}`);
          }

          const fileContent = fs.readFileSync(filePath, "utf8");
          const fileName = path.basename(filePath);

          const tempProject = new Project({
            useInMemoryFileSystem: true,
            compilerOptions: {
              allowJs: true,
              allowSyntheticDefaultImports: true,
            },
          });

          // Add file content to in-memory project
          sourceFile = tempProject.createSourceFile(fileName, fileContent);
          if (process.env.SUNLINT_DEBUG)
            console.log(
              `ðŸ”§ [S020] Temporary project created successfully with file: ${fileName}`
            );
        } catch (projectError) {
          if (process.env.SUNLINT_DEBUG)
            console.log(
              `ðŸ”§ [S020] Failed to create temporary project:`,
              projectError.message
            );
          throw projectError;
        }
      }

      if (sourceFile) {
        const symbolViolations = await this.symbolAnalyzer.analyze(
          sourceFile,
          filePath
        );
        symbolViolations.forEach((v) => {
          const key = `${v.line}:${v.column}:${v.message}`;
          if (!violationMap.has(key)) violationMap.set(key, v);
        });
        if (process.env.SUNLINT_DEBUG)
          console.log(
            `ðŸ”§ [S020] Symbol analysis completed: ${symbolViolations.length} violations`
          );
      } else {
        if (process.env.SUNLINT_DEBUG)
          console.log(
            `ðŸ”§ [S020] No source file found, skipping symbol analysis`
          );
      }
    } catch (error) {
      console.warn(`âš  [S020] Symbol analysis failed:`, error.message);
    }

    const finalViolations = Array.from(violationMap.values()).map((v) => ({
      ...v,
      filePath,
      file: filePath,
    }));
    if (process.env.SUNLINT_DEBUG)
      console.log(
        `ðŸ”§ [S020] File analysis completed: ${finalViolations.length} unique violations`
      );
    return finalViolations;
  }

  cleanup() {
    if (this.symbolAnalyzer?.cleanup) this.symbolAnalyzer.cleanup();
  }
}

module.exports = S020Analyzer;
