// @sun-asterisk/sunlint/rules/security/S055_content_type_validation/analyzer.js
/**
 * S055 Main Analyzer - Content-Type Validation in REST Services
 * Primary: Symbol-based analysis (when available)
 * Fallback: Regex-based for all other cases
 * Purpose: Detect REST endpoints that process request body without validating Content-Type
 * Command: node cli.js --rule=S055 --input=examples/rule-test-fixtures/rules/S055_content_type_validation --engine=heuristic --verbose
 */

const S055SymbolBasedAnalyzer = require("./symbol-based-analyzer.js");

class S055Analyzer {
  constructor(options = {}) {
    if (process.env.SUNLINT_DEBUG) {
      console.log(`🔧 [S055] Constructor called with options:`, !!options);
      console.log(
        `🔧 [S055] Options type:`,
        typeof options,
        Object.keys(options || {})
      );
    }

    this.ruleId = "S055";
    this.ruleName = "Do not expose version information in response headers";
    this.description =
      "Detect REST endpoints that process request body without validating Content-Type";
    this.semanticEngine = options.semanticEngine || null;
    this.verbose = options.verbose || false;

    this.config = {
      useSymbolBased: true,
      fallbackToRegex: false,
      regexBasedOnly: false,
      prioritizeSymbolic: true, // Prefer symbol-based when available
      fallbackToSymbol: true, // Allow symbol analysis even without semantic engine
    };

    try {
      this.symbolAnalyzer = new S055SymbolBasedAnalyzer(this.semanticEngine);
      if (process.env.SUNLINT_DEBUG)
        console.log(`🔧 [S055] Symbol analyzer created successfully`);
    } catch (error) {
      console.error(`🔧 [S055] Error creating symbol analyzer:`, error);
    }
  }

  async initialize(semanticEngine = null) {
    if (semanticEngine) {
      this.semanticEngine = semanticEngine;
    }
    this.verbose = semanticEngine?.verbose || false;

    // Initialize both analyzers
    await this.symbolAnalyzer.initialize(semanticEngine);

    // Ensure verbose flag is propagated
    this.symbolAnalyzer.verbose = this.verbose;

    if (this.verbose) {
      console.log(`🔧 [S055 Hybrid] Analyzer initialized - verbose: ${this.verbose}`);
    }
  }

  analyzeSingle(filePath, options = {}) {
    if (process.env.SUNLINT_DEBUG)
      console.log(`🔍 [S055] analyzeSingle() called for: ${filePath}`);
    return this.analyze([filePath], "typescript", options);
  }

  async analyze(files, language, options = {}) {
    if (process.env.SUNLINT_DEBUG) {
      console.log(`🔧 [S055] analyze() method called with ${files.length} files, language: ${language}`);
    }

    const violations = [];

    for (const filePath of files) {
      try {
        if (process.env.SUNLINT_DEBUG) {
          console.log(`🔧 [S055] Processing file: ${filePath}`);
        }

        const fileViolations = await this.analyzeFile(filePath, options);
        violations.push(...fileViolations);

        if (process.env.SUNLINT_DEBUG) {
          console.log(`🔧 [S055] File ${filePath}: Found ${fileViolations.length} violations`);
        }
      } catch (error) {
        console.warn(`❌ [S055] Analysis failed for ${filePath}:`, error.message);
      }
    }

    if (process.env.SUNLINT_DEBUG) {
      console.log(`🔧 [S055] Total violations found: ${violations.length}`);
    }

    return violations;
  }

  async analyzeFile(filePath, options = {}) {
    if (process.env.SUNLINT_DEBUG) {
      console.log(`🔧 [S055] analyzeFile() called for: ${filePath}`);
    }

    // 1. Try Symbol-based analysis first (primary)
    if (this.config.useSymbolBased &&
        this.semanticEngine?.project &&
        this.semanticEngine?.initialized) {
      try {
        if (process.env.SUNLINT_DEBUG) {
          console.log(`🔧 [S055] Trying symbol-based analysis...`);
        }
        const sourceFile = this.semanticEngine.project.getSourceFile(filePath);
        if (sourceFile) {
          if (process.env.SUNLINT_DEBUG) {
            console.log(`🔧 [S055] Source file found, analyzing with symbol-based...`);
          }
          const violations = await this.symbolAnalyzer.analyzeFileWithSymbols(filePath, { ...options, verbose: options.verbose });

          // Mark violations with analysis strategy
          violations.forEach(v => v.analysisStrategy = 'symbol-based');

          if (process.env.SUNLINT_DEBUG) {
            console.log(`✅ [S055] Symbol-based analysis: ${violations.length} violations`);
          }
          return violations; // Return even if 0 violations - symbol analysis completed successfully
        } else {
          if (process.env.SUNLINT_DEBUG) {
            console.log(`⚠️ [S055] Source file not found in project`);
          }
        }
      } catch (error) {
        console.warn(`⚠️ [S055] Symbol analysis failed: ${error.message}`);
        // Continue to fallback
      }
    } else {
      if (process.env.SUNLINT_DEBUG) {
        console.log(`🔄 [S055] Symbol analysis conditions check:`);
        console.log(`  - useSymbolBased: ${this.config.useSymbolBased}`);
        console.log(`  - semanticEngine: ${!!this.semanticEngine}`);
        console.log(`  - semanticEngine.project: ${!!this.semanticEngine?.project}`);
        console.log(`  - semanticEngine.initialized: ${this.semanticEngine?.initialized}`);
        console.log(`🔄 [S055] Symbol analysis unavailable, using regex fallback`);
      }
    }

    if (options?.verbose) {
      console.log(`🔧 [S055] No analysis methods succeeded, returning empty`);
    }
    return [];
  }
}

module.exports = S055Analyzer;
